### ISSUE TITLE
File Management

### ISSUE THREAD

**Issue by spacekookie** [rust-lang-nursery/cli-wg#7]
In the first meeting, we [discussed](https://github.com/rust-lang-nursery/cli-wg/blob/master/logs/2018-02-20-cli-wg-initial-meeting.log#L74) the task of file management on different platforms (particularly configurations) and how it can be made better.

@Eijebong summarised it like this

> If I need a config file, I don't want to know that it should be `${XDG_CONFIG:${XDG_HOME}/.config:/home/${user}/.config` on linux, `%AppDir%/App/` on windows and something else on osx [...]

There is a crate for "determing system configuration" ([app-dirs-rs](https://github.com/AndyBarron/app-dirs-rs)) but it seems unmaintained and not up to date

**Comment by killercup** [comment]
The [configure](https://github.com/withoutboats/configure) crate by @withoutboats abstract about configuration in general (as the name suggests), but doesn't seem to have an adapter for configuration _files_ right now (except for Cargo.toml, but that's not the use case we have in mind I guess).

**Comment by kbknapp** [comment]
I think it'd good to lay out some requirements for a crate(s) that would fill this gap:

 * Accepts a file name to either load/save
 * Abstracts away platform specific paths (i.e. as the *application* developer I only want to say, "myapp.conf" and have the crate handle *where* and *how* "myapp.conf" gets loaded/saved

I'd be OK with having multiple sub crates for various platforms, and then a "parent" crate that allows abstracting over the platform where all I do is specify a file name I want to load/save.

**Comment by kbknapp** [comment]
Bonus if the crate can allow me to search custom directories, or tweak the order (on platforms where applicable).

**Comment by withoutboats** [comment]
> but doesn't seem to have an adapter for configuration files right now (except for Cargo.toml, but that's not the use case we have in mind I guess).

Nope, the configure crate's default "source" is definitely designed for use cases where the person configuring the application is also the author - such as network services. However, the intent is for *libraries* to use configure, so that the application author can have total control over the source of configuration.

A configuration source that integrates with configure and is designed for CLIs would be a great addition, and possibly one I'd be interested in upstreaming into configure proper.

**Comment by yoshuawuyts** [comment]
I'm thinking this issue might be part of a bigger topic.
- How do we store __user-created configuration files__ cross-platform? (So far this issue has only touched on this one specifically. E.g. Linux `$HOME/.config`)
- Where do we store __temporary files__ cross-platform? (Does not need to persist between boots, e.g. Linux `/tmp`.).
- How do we store __essential user data files__ cross-platform? (Stuff like login tokens, usually generated by applications. E.g. `$HOME/.local` on Linux.)
- How do we store __non-essential data files__ cross-platform? (Needs to persist between boots, e.g. Linux `$HOME/.cache`.)

It's probably uncommon for a single application to use all of these, but one or more should be common enough. It feels like these questions are part of the same problem; perhaps it might be useful to consider all of these questions as part of this discussion?

**Comment by spacekookie** [comment]
For the question of **temporary files** there is already [a crate](https://crates.io/crates/tempdir) which seems to do its job quite nicely (though I've only used it in limited scenarios so far, maybe it can be improved!)

As for the rest…I think it would be pretty cool if we could create (or find and improve existing) crates that mirror the same behaviour for other configuration, essential and non-essential data files as well.

It should be as simple as saying `ConfigDir::new("my_app_name")` and being able to write and read configurations from it.

**Comment by killercup** [comment]
Hi, @soc! The Rust CLI working group is talking about cross-platform configuration file management and your _directories_ crate has come up. Looking at your Github profile, I see you have a Java _directories_ package as well, so you seem have some expertise in this area. Wanna chime in here? :)

**Comment by soc** [comment]
@killercup Sure, how can I help?

**Comment by killercup** [comment]
@soc awesome! We were currently doing some research about the status quo of crates that are useful when writing CLI tools, work cross-platform and are maintained. For example, we want to come up with a good story around how to easily configure a CLI tool—with config files, env vars, and CLI flags. This issue is focussing on the handling of config files. @kbknapp already listed some good requirements in <https://github.com/rust-lang-nursery/cli-wg/issues/7#issuecomment-367085114>.

Do you think _directories_ is a good foundation here? What are your plans for it? Can we help you get it to 1.0? :)

(@spacekookie and @yoshuawuyts probably have more to say!)

**Comment by soc** [comment]
> For example, we want to come up with a good story around how to easily configure a CLI tool—with config files, env vars, and CLI flags.

_directories_ is intentionally focused solely on dealing with operating system defaults.
The reasoning for this is not because I believe that other venues for configuration are not important, but to provide the most minimal, stable API I can get away with.

For instance when dealing with CLI flags, the first issue you have is that of style (`-h` and `--help` vs. `-help` etc.) There is potentially a lot of complexity involved when trying to provide an CLI interface that makes everyone happy.

I do think that the library is a good foundation for dealing with the operating system standards part of your goals.

I believe that dealing with CLI flags should probably be done in a separate library, or in a way more specific to the individual application's needs, because dealing with CLI flags is very application-specific.

In the end individual applications already need to have some custom code anyway to deal with migrating from storing their data directly in `$HOME` to following the platform standards. Dealing with CLI flags will probably be the same.

That's why _directories_ does not get involved with creating directories itself, or checking for their existence etc. ... _directories_ only tells developers which directories they should be using, because I know of the complexities involved of making (legacy) applications actually use them without breaking things for existing users.

> What are your plans for it? Can we help you get it to 1.0? :)

My plan is to declare it as stable as fast as possible. I think the main blockers are

- having more people use and test it, to make sure it works
- a thorough review of the decisions made concerning the various paths chosen
- a review of the Windows-specific code

**Comment by soc** [comment]
I have created tickets for the remaining issues I mentioned: https://github.com/soc/directories-rs/issues/1

**Comment by soc** [comment]
A more general note: There is a vast difference between selecting and standardizing on crates that provide certain functionality (like CLI parsing, config file parsing) and having one standardized way of handling application configuration:

With the former you probably get crates that do almost everything and allow configuration of almost everything.

With the latter, you want to be highly selective and make actual choices how things can be specified, and not allow a free for all in terms of configuration.

**Comment by killercup** [comment]
As you've noticed, I've opened some issues at directories-rs. I'd hold off on releasing a 1.0 before there are some consumers of the crate.

- - -

> There is a vast difference between selecting and standardizing on crates that provide certain functionality (like CLI parsing, config file parsing) and having one standardized way of handling application configuration

Absolutely. We already have some _great_ libraries for CLI args, and I'd love to have an equally as good story for dealing with config files. That is not one crate – it's several build on top of and complementing each other :)

(We'll hopefully see more concrete proposals for this in #6!)

**Comment by spacekookie** [comment]
I think the focus should be less on a config file *format* and more on an API to get to those files. As a developer I might still want to be able to chose a format, say `json` or `toml` or `ini` via whatever serde backend exists to read/ write my configuration files. But I don't want to have to worry about where to put it.

Not sure why you brought up CLI parsing. Although thinking about it now, I'm not sure how clap.rs handles windows arguments :sweat_smile:

**Comment by killercup** [comment]
> Not sure why you brought up CLI parsing.

I brought it up, sorry :)

So, I've been thinking about what an all-around config solution might look like. **We should not implement such a thing right now, but discuss what needs to happen to get there!**

- - -

Here's a small proposal that integration ideas from clap (v3, this is future!) and configure to get the discussion going:

```rust
#[derive(Debug, Deserialize, Clap, Configure)]
#[config(prefix = "diesel")]
struct Args {
    #[clap(short = "q", long = "quiet")]
    quiet: bool,
    #[clap(long = "database-url")]
    database_url: bool,
    #[clap(subcommands)]
    command: DieselCliCommand, // an enum defining subcommands with their own fields and attributes
}

fn main() {
    let args = Args::configure()
        .read_from(configure::adaptors::config_file::toml("diesel_cli.toml")) // Invokes serde
        .read_from(configure::adaptors::env_file()) //  dotenv
        .read_from(configure::adaptors::env()) // std::env
        .read_from(configure::adaptors::clap_auto_init()); // Clap incl. early exit on `-h` and stuff like that
}
```

You can then:

- pass `--database-url=something.sqlite`
- execute the program with `env DIESEL_DATABASE_URL=something.sqlite`
- Have a `.env` file with `DIESEL_DATABASE_URL=something.sqlite`
- Have a `~/.config/diesel_cli.toml` file


- - -

Is that approximately the direction in which you want to go? What needs to happen to get there?

**Comment by TeXitoi** [comment]
I think the CLI/conf/env story should be in another issue.

**Comment by killercup** [comment]
Sure, that was just for inspiration and to set some context. (If you have other use cases/ideas, please tell us :))

**Comment by TeXitoi** [comment]
I have a couple of request in the structopt issues about that (no ideas, but persons wanting something like https://github.com/rust-lang-nursery/cli-wg/issues/7#issuecomment-367673115)

**Comment by kbknapp** [comment]
Like @spacekookie said, I think it should focus on abstracting over platform specific issues and not on the format, or providing "key->value" style API.

As the application writer, I want to just specify a file name, and let this crate handle *where* to store it.

Then later on  someone could write a generic crate to abstract over this configure crate, using something like serde to give a key->value style API.

Here's how I see the crate structure playing out (note, the crate names are just generic and not referring to anything existing right now).

![config](https://clap.rs/config.png)

**Comment by Screwtapello** [comment]
At a former employer, I wrote a config file management library (in Python) that turned out to be popular with my fellow developers (because it was easy to add to an existing project) and with our operations staff (because all our tools worked the same way, and the configuration was flexible enough for most of our use-cases). It worked like this:

- an application would include a configuration file containing all the generic defaults, in the Python `ConfigParser` format (basically, an INI file)
- the application calls into the library, passing the application name and the defaults file
- the library reads the defaults
- the library reads `/etc/xdg/$appname/config.cfg` (the standard XDG system-wide config directory, if it exists) and overlays those settings on top of the defaults
- the library reads `~/.config/$appname/config.cfg` (the standard XDG per-user config directory, if it exists` and overlays those settings on top of the defaults
- For each section and key in the combined configuration, the library would build a string like `$appname_$section_$key` and upper-case it. If an environment variable with that name existed, its value would replace the value loaded from the config files
- the library returns the fully-populated configuration data to the application

Pros:

- The application only needs to call a single function.
- The defaults file can contain all the possible config settings, example values and even documentation for them
    - maybe not the best possible location for such things, but still better than "scattered across the application in each bit of code that reads a config variable"
- System-wide configuration is useful for provisioning tools like Ansible/Puppet/Chef/etc. that deploy the tool to automatically configure it to work on that host (for example, setting a default HTTP proxy, or picking a geographically close rendezvous server)
- Per-user configuration means users can set things up the way they like them
- Environment variable configuration means one tool can launch another tool and force some configuration option it needs (like output format, or log file name)
- At each stage, you can override *some* defaults without replacing *all* of them
    - for example, you can use Ansible/Puppet/Chef/etc. to change the "default HTTP proxy" in the system-wide config without worrying about users who have created their own config missing out
- Because the INI format is limited, overlaying configuration files is simple, with predictable results

Cons:

- It can be difficult to fit an application's configuration needs into the limited vocabulary of INI files
- With no configuration schema, the application has to do all the deserialization/validation work itself
- If a config section `foo` has key `bar_baz`, and section `foo_bar` has key `baz`, both will be mapped to the environment variable `$appname_FOO_BAR_BAZ` and there isn't really any way around that.
- There's no good way to extend this to CLI parsing, not least because it would be impossible to generate decent `--help` output from the limited information we have

If I were to attempt something similar in Rust:
- I'd try to find a way to build it around a configuration schema object (following the model of `serde` and `structopt`) instead of tossing around raw config files
- I'd probably leave the environment-variable config the same; it's clunky but we didn't need it very often, so it wasn't a problem in practice
- I'd absolutely 100% require the ability to overlay config files for different sources, though—I don't know how that would work exactly with richer data formats than INI, but I'd have to find a way

**Comment by Screwtapello** [comment]
> I think the focus should be less on a config file *format* and more on an API to get to those files.

One reason to consider a standard config file format, or at least a standard config data model: on Windows, perhaps the standard configuration source could/should be the Registry, rather than the filesystem?

**Comment by soc** [comment]
After some research on that, it seems that most developers recommend and prefer files over the registry:
- https://softwareengineering.stackexchange.com/questions/144238/ini-files-or-registry-or-personal-files
- https://blog.codinghorror.com/was-the-windows-registry-a-good-idea/

**Comment by derekdreery** [comment]
Since this thread is about the *location* of config files rather than their *contents* this may be a bit off topic, but here goes anyway:

Similar to how structopt works, I'd love to do

```rust
#[derive(Structconfig)]
pub struct Config {
    timeout: u8
    #[structconfig(name="retries", default=3)]
    no_of_retries: u8,
    files: Vec<PathBuf>,
}
```
and have all the config stuff taken care of for me!

**Comment by derekdreery** [comment]
@Screwtapello 

How did your code deal with first run, if there wasn't a config file? Did it assume you wanted to use the defaults, or did it exit and prompt you to create a config file? (or did it walk you through creating the config file interactively?)

**Comment by Screwtapello** [comment]
@derekdreery 

At first run, it would use the defaults. For the various tools we created, every config option always had a sensible out-of-the-box default. Things the program absolutely could not know without asking would generally be command-line arguments, not config options.

It's a big world, and I'm sure there's some potential config options that cannot possibly have a sensible default, but I can't think of one right now. If anyone has an example, I'd love to hear it.

**Comment by epage** [comment]
Another aspect of config management to consider is passwords.  Looks like there is a [`keyring`](https://github.com/hwchen/keyring-rs) crate that could use some polish and advertising.

**Comment by kalefranz** [comment]
Hey everyone. I'm the current dev lead of conda, which is a cross-platform, system-level package manager.  Currently written in python--but we're in the initial stages of considering transitioning key pieces to rust.

Just wanted to add to this discussion how we do configuration, because it's been powerful and has worked out very well.  It's also very similar to what @Screwtapello described.

For each invocation of our executable, we build up a configuration context object from four sources of configuration information:

1. hard-coded default values
2. (potentially multiple) configuration files, including support for files in "`.d`" directories
3. environment variables
4. command line flags

These are linearized in a way that the configuration sources conceptually closest to the process invocation take precedence.  That is, if a configuration parameter is provided as a CLI flag, but also provided in a configuration file, the CLI-provided value would win.  I guess the insight here is that most CLI applications deal with at least one configuration file, environment variables, and CLI flags anyway, and we've just realized that they all represent basically the same type of information, and can be generalized and unified.

One capability that was especially important for us to add was the ability for sysadmins to lock down configuration for the entire system in "lower-level" read-only files.  As we merge the sources of configuration information, we provide a flag sort of like the css `!important` that lets the lower-level value be the final value.

I don't want to go into too much detail here.  There's a [blog post](https://www.anaconda.com/blog/developer-blog/conda-configuration-engine-power-users) with more details, including how we deal with merging sequence and map-type configuration parameters.  I did want to point all this out though as support for the usefulness of what @Screwtapello described.

**Comment by Screwtapello** [comment]
>  As we merge the sources of configuration information, we provide a flag sort of like the css !important that lets the lower-level value be the final value.

An alternative model that achieves the same goal is to have separate "config" and "override" files:

* system-wide config
    * per-user config
        * environment variables and command-line flags
    * per-user overrides
* system-wide overrides

The advantage over an `!important` flag is that you don't need special syntax in your config-file format (and therefore `serde`, etc.) while the disadvantage is that you have nearly twice as many config locations to document, and for users to check when diagnosing surprising behaviour.

**Comment by derekdreery** [comment]
> the disadvantage is that you have nearly twice as many config locations to document, and for users to check when diagnosing surprising behaviour

@Screwtapello you could mitigate this by being able to generate something like the following

```
# Configs - some config option
 1. There was no value at system-wide level. *value = default*
 2. Found value *newvalue* at user level *value = newvalue*
 3. There was no value at env/cli level *value = newvalue*
 4. There was no value at user override level *value = newvalue*
 5. Found value *newvalue2* at system override level *value = newvalue2*
 6. Final value for *config option* is *newvalue2*
```
```

**Comment by derekdreery** [comment]
> the disadvantage is that you have nearly twice as many config locations to document, and for users to check when diagnosing surprising behaviour

@Screwtapello you could mitigate this by being able to generate something like the following

```
# Configs - some config option
 1. There was no value at system-wide level. *value = default*
 2. Found value *newvalue* at user level *value = newvalue*
 3. There was no value at env/cli level *value = newvalue*
 4. There was no value at user override level *value = newvalue*
 5. Found value *newvalue2* at system override level *value = newvalue2*
 6. Final value for *config option* is *newvalue2*
```

**Comment by soc** [comment]
One option would be to have an API like

    Config::from(system_overrides, commandline, environment, config_file, legacy_config_file, system)

Where people describe the order of settings they want to have and the library resolves settings in that order until a value is found.

I think an additional bit that's important to get right is to track the origin of each setting, so that people don't end up with `some_setting = "value"`, but `some_setting = ("value", source)` I think this would make it way more transparent to understand and debug where settings come from.

**Comment by Screwtapello** [comment]
> I believe having some hard-coded, common-sense lookup scheme would be nice, but I fear that many applications would not fit well into it.

Do you have particular examples of applications that would not fit well? This lookup scheme is a superset of the lookup schemes of POSIX-tradition applications I can think of offhand, and resembles what I know of other platforms (macOS has system-wide `/Library` versus per-user `~/Library`, Windows has `HKEY_LOCAL_MACHINE\Software` versus `HKEY_CURRENT_USER\Software`. Sure, not every existing piece of software follows exactly that scheme, but I imagine that's largely because they had to implement it themselves from scratch; if I could get that scheme for free from a library, I certainly would.

> I think an additional bit that's important to get right is to track the origin of each setting, so that people don't end up with `some_setting = "value"`, but `some_setting = ("value", source)`

I would love a library that provided a citation for every config setting (or every piece of input data in general), but as far as I know `serde` likes to throw away all that information once the deserialization is done. An alternative would be to make an opinionated library that supports only a single config file format, but tracks citations... which is doable, but would certainly be less popular.

**Comment by soc** [comment]
My concern is that every existing applications is probably littered with special cases, migration schemes and special rules how settings are handled. I'm not sure it is possible to encode all this into an API without making it extremely hard to use.

**Comment by luser** [comment]
> There is a crate for "determing system configuration" (app-dirs-rs) but it seems unmaintained and not up to date

I've used `app-dirs` in a few projects and it was very close to what I want. It's unfortunate that it's unmaintained, but we should crib ideas from it for a replacement, if nothing else!

**Comment by soc** [comment]
@luser Is there anything in `app-dirs` that you feel is missing in `directories` or is done better?

**Comment by spacekookie** [comment]
@soc I think building a comprehensive API that handles config files in the (what's considered for that platform to be) "best practise" is the exact way you fix those corner cases though. A lot of people always assume they know it better or want to have configs somewhere else because…<weird reasons>. Making it easier to follow the de facto *standard way* will reduce the amount of applications that don't do it that way, essentially getting rid of the problem :wink:

**Comment by soc** [comment]
@spacekookie I totally agree on making it easier to follow the standard!

My concern is that if you build an API that deals with all the corner cases you are pretty much making the API worse for the "good citizens" of the ecosystem, e. g. an API that says

> ok, here is where the config dir should be, except when we have a legacy dir over here, or some custom settings over there etc.

is very different from an API that says

> give me the name of your app, and I'll tell you the paths where you should store things

**Comment by killercup** [comment]
I can totally imagine that the just-give-me-a-config crate that internally uses all the best practices and handle common stuff is just a bunch of foundational crates used in a specific way. So, if you wanted to customize it, you should be able to do what the JGMAC crate (need to work on an abbreviation) does (by doing a good amount of copy-pasta). Maybe we can even write the JGMAC crate in a style that makes "ejecting" easy.

This is an easier to maintain system than trying to do provide a super customizable API. You can think of it like a quicli for configs.

**Comment by whmountains** [comment]
IMO the Javascript CLI ecosystem really get's this right.  See https://github.com/davidtheclark/cosmiconfig

**Comment by luser** [comment]
> @luser Is there anything in app-dirs that you feel is missing in directories or is done better?

`directories` seems very sensible and looks like it has roughly everything I would want. My only quibble would be that it does not ensure the directories exist, but that's not a deal-breaker. Getting `directories` to a 1.0 release would be a good thing for this working group to push on.

**Comment by soc** [comment]
Yes, that was an explicit design decision, because the decision to do when something doesn't exist is extremely application-specific.

For instance, if you run an application for the first time, the directory doesn't exist, and you absolutely want the application to create it.

But for instance if you have an application that uses a legacy .app folder instead of following XDG, you probably want to keep using that old folder as-is and not create the new directories.

**Comment by luser** [comment]
I've been using `structopt` in a few projects lately and I'd love to see a config file solution that works as nicely as it does. Maybe building on withoutboats' `configure` crate, something like:
```rust
#[derive(Deserialize, Configure, Default)]
struct Config {
 //...
}

fn work() -> Result<(), Error> {
  let config: Config = read_config()?;
}
```

Where `read_config` would use `directories` or something under the hood to locate an appropriate config file, and if not present use `Default::default()`.

Some other things it'd be nice to consider:
* Good error messages for malformed config files. I know @geal has been looking into this recently: https://github.com/serde-rs/serde/issues/1184
* Potentially having a way to generate a config file containing defaults would be nice.

**Comment by i30817** [comment]
Just butting in with my uninformed opinion:

Not only in .cfg files but even in scripts and custom cli environments (dosbox for example) programmers will try to get the user to insert Paths in user readable and writable format. As soon as users try to write this to file, this will never work portably without major sacrifices to correctness, which leads to 'oh you want a portable app with data pre-configured? You can do that, but you have to edit the .cfg/script to change the path separator'.

`relative_from_cfg(path: &str) -> PathBuff`

For my preferred 'major sacrifice to correctness' I'd like a method to read a relative path that *assumes* a single and only path separator for all platforms (/ because it's a forbidden filename character in both windows and unix) and return Mone if it has even the possibility of not being relative (start with /. \\ or [A-z]+: etc). I
'm aware that unix relative paths can start with c: , but if you want to bring sanity to the 'user made portable paths' you can't accept everything possible. Document that it's clearly for user made paths and leave it at that.

**Comment by epage** [comment]
@i30817 is your concern with reading/writing a config file on the same system or sharing config files across systems?

If sharing across systems, can you give more concrete details so we can flesh out what the requirements are for this?

**Comment by i30817** [comment]
sharing across systems. Common usecase is placing files which contain relative paths to some config dir (of a portable app or app that can be made portable like firefox, or dolphin or retroarch).

Supposedly users can use 'a/relative/path' syntax in windows. In practice, nearly all C and C++ programs blow up with this kind of relative path because they only `ifdef _WIN32 '\\' else '/'` etc.

There is the complication that a 'c:/a/relative/path' is valid unix syntax for a relative path, but on windows it's a absolute legal path (and maybe illegal if you try to join it to another absolute path), so such a reader function should validate input to see if the user really put in something portable.

**Comment by Dylan-DPC** [comment]
We have opened a ["RFC-lite"](https://github.com/rust-clique/confy/pull/2) on confy with reference to this. Do leave your feedback.

**Comment by sharkdp** [comment]
The `directories` crate has been mentioned a few times in this ticket. I have recently started to use it in one of my command-line tools and now this issue came up: https://github.com/sharkdp/bat/issues/151. Do people here have any opinion on this? Should *command-line tools* on macOS use `~/Library/Preferences` or `~/.config`?

**Comment by soc** [comment]
@i30817 Interesting, I played with exactly this idea a few weeks ago:

- Have one standard serialization format used to read/write paths from/to config riles
- Rule out "dangerous" characters in paths
- Provide absolute path placeholders (e. g. you can say "this is an absolute path starting from e. g. the OS' config directory" and have it return the appropriate absolute directory for each platform) so that one can share config files with paths across operating systems.

**Comment by i30817** [comment]
Yeah, well i reconsidered point 2 myself. I don't mind it being a warning during development, but i don't think it should be forced. Especially since most cli config paths are not controlled by devs but by the user and sometimes they themselves don't control what another tool/user named the files (imagine a file database that is using utf-8).

**Comment by soc** [comment]
Yeah ... my approach comes with the expectation that developers are in control of their config file names. I think that's a reasonable assumption in 90% of the cases.

In general, I think there is no having the cake and eating it: either you rule out the various OS' filesystem craziness, or you allow it and have the complexity go through the roof. 

I think there is value in having some crate that says "if you don't depend on the crazy parts of Windows path handling, this crate makes your life way easier".

**Comment by i30817** [comment]
It's true that is a concern more of files that interact more with user paths that the typical config file (though 'last used recent path' is a frequent exception), but i'm thinking that point 1 and 3 are sufficiently valuable that you want to give the ability for them to be used in those paths.

**Comment by soc** [comment]
@sharkdp It would be more correct to use Apple's plist format if you write configuration to ~/Library/Preferences, but that's why the API avoids making any judgement on _what_ you do with the paths _directories_ gives you.

If I have some time in the future, I'd love to propose some RFC for Redox that overhauls config handling there and extend  _directories_ to support it.

**Comment by pickfire** [comment]
https://crates.io/crates/dirs looks like a good choice for me even with it's limited features but the API is a lot simpler compared to https://crates.io/crates/directories.

**Comment by soc** [comment]
Yes, that's intentional, _dirs_ just exposes individual functions, but if you want to derive paths for applications then you should absolutely use _directories_' `ProjectDirs` instead.

**Comment by kbd** [comment]
> Should *command-line tools* on macOS use `~/Library/Preferences` or `~/.config`?

I would strongly expect any command line tool to use XDG standard location for config. There is no command line tool I know of that stores its config in `~/Library/Preferences`. I know I'm not alone in source-controlling my dotfiles in `~/.config` -- while not source-controlling anything in `~/Library/Preferences` -- so storing config in `~/Library/Preferences` makes that less convenient.

Besides the lack of consistency in storing things in one location on Linux and another on Mac, if Rust utilities alone start storing things in `~/Library/Preferences` it's going to be a Java Swing-style uncanny-valley sign of "this is a Rust app".

**Comment by soc** [comment]
@kbd As a bit of data: `dirs` and `directories` have been out for a while, and I have received no complaints about those libraries adhering to macOS conventions, at all.

**Comment by lavifb** [comment]
I would like to second @kbd. Command line tools are often cross-platform for *nix and as a result the config files are stored similarly to the way they do on Linux, at `~/.config`. It doesn't make much sense to store them in a harder to reach and less *nix-y place like `~/Library/Preferences`.

**Comment by reitermarkus** [comment]
I can only confirm what @kbd and @lavifb are saying. I have yet to see a CLI tool which uses `~/Library/Preferences` for configuration files.

**Comment by bitwalker** [comment]
Just wanted to chime in with my two cents here as well, this actually caught me off guard recently with a tool I was using, `llvmenv`, which I was making some changes to. It instructed me (and I would've expected no different) to put files in `$XDG_CONFIG_HOME/llvmenv`, and that a variety of other associated files (cache, data) would be put in other XDG directories accordingly. I did so, but the tool wasn't picking up my config, and I assumed it was a bug of some kind. It was only after I dug through llvmenv, traced through to `dirs`, that I realized what was going on. It is highly unusual in my experience to find command line tooling _with support for XDG_ that does not honor it on macOS - so unusual that I don't think the author of that tool even realized this was happening on macOS.

If the default directories end up in `~/Library` or wherever, then I think it should be a convention that if the `XDG_*` env vars are set, that those directories are used instead. If they are set, that is a strong indication that the user expects those paths to be used in place of whatever other conventions the OS uses.

I basically never assume configs or other data associated with my shell environment are going to end up anywhere other than in the XDG base directories I have set, or in the worst case, my home directory - but I would never even think to look in `~/Library/Preferences` for that kind of thing.

For tooling I build, I certainly will be overriding this behaviour in `dirs` or any other such library I use so that XDG is always honored first.

**Comment by soc** [comment]
I closed https://github.com/soc/directories-rs/issues/47, please have a look at the explanation in https://github.com/soc/directories-rs/issues/47#issuecomment-478337412 for details.

**Comment by sharkdp** [comment]
@soc I understand that you do not want to include this in directories-rs, which probably also targets other (GUI) applications.

However, as we have heard from multiple people in this thread (and multiple people in `bat`s issue tracker), the official standard (`~/Library/Preferences`) on MacOS does not seem to be followed by any *command line* applications, whereas the XDG-standard seems to be used quite a lot. In this case, I'd rather listen to the actual users and follow what other command-line options do, instead of following the "official standard".

**Comment by lavifb** [comment]
Also I'm not sure I understand the cache problem. Is it a big problem that cached files for CLI apps might be backed up? CLI caches do not tend to be that big in my experience.

But even if it were a big problem, this is not a reason config files can't be placed in `~/.config`. People don't manually look at or edit caches very often. Config files, on the other hand, are very important _especially_ in CLI applications. There is often no other way to configure these apps. Placing these files in a convenient and generally agreed upon location for users shouldn't be dismissed out of hand.

**Comment by soc** [comment]
@sharkdp 

> However, as we have heard from multiple people in this thread (and multiple people in bats issue tracker), the official standard (~/Library/Preferences) on MacOS does not seem to be followed by any command line applications, whereas the XDG-standard seems to be used quite a lot. In this case, I'd rather listen to the actual users and follow what other command-line applications do, instead of following the "official standard".

The difference between Linux and macOS is that Linux is a system written by developers, for developers.

On the other hand., macOS is built by Apple, owned by Apple and controlled by Apple. The opinions of developers frankly doesn't matter on macOS – they are not calling the shots. If Apple ever decides to clamp down on applications not following their rules, I'm not going to take responsibility for the ensuing breakage.

Just as a reminder,

> does not seem to be followed by any command line applications

is an excuse I heard more often than I care to count on Linux, too. `directories-rs` does exactly what it says on the tin: it follows the platforms standards.

**Comment by soc** [comment]
@lavifb 

>  Is it a big problem that cached files for CLI apps might be backed up? CLI caches do not tend to be that big in my experience.

Pretty much every dependency manager on this planet is a CLI. Have a look at the cached stuff of Maven, npm or cargo. The cache directories are _huge_.

**Comment by lavifb** [comment]
> @lavifb
> 
> > Is it a big problem that cached files for CLI apps might be backed up? CLI caches do not tend to be that big in my experience.
> 
> Pretty much every dependency manager on this planet is a CLI app. Have a look at the caches of Maven, npm or cargo. The cache directories are _huge_.

But all of those store cache in the home dir on MacOS: 
- [`~/.npm/`](https://docs.npmjs.com/files/folders.html#cache)
- [`~/.cargo/`](https://doc.rust-lang.org/cargo/commands/index.html)
- [`~/.m2/`](https://maven.apache.org/guides/mini/guide-configuring-maven.html)

More importantly, their config files are not buried somewhere deep in preferences but are all easily accessible in the home dir.

**Comment by soc** [comment]
Yes, I hope you now understand the issue with that (on both Linux and macOS).

**Comment by lavifb** [comment]
@soc
I think we just have different philosophies for this kind of library. Thanks for explaining your reasoning


@sharkdp
> This is why I feel that there is a need for a `directories-rs`-like library that uses the XDG-standard on both Linux/MacOS and the "Known Folders" on Windows.

I do, however, think there is room for another crate that makes a different tradeoff here.

**Comment by sharkdp** [comment]
> > This is why I feel that there is a need for a `directories-rs`-like library that uses the XDG-standard on both Linux/MacOS and the "Known Folders" on Windows.
> 
> I do, however, think there is room for another crate that makes a different tradeoff here.

Absolutely. That's why I said that I completely understand that @soc does not want to change `directories-rs`.

**Comment by soc** [comment]
Alternatively we could encourage developers to write applications which are well-behaved citizens of the platform they run on.

In the beginning of medicine people also thought there was room for not washing hands before a procedure. Thankfully, washing hands became non-optional, even though many considered it to be surprising and unexpected at that time.

I'm optimistic that – with an effort to educate developers – a similar success story can be achieved here.

**Comment by kbd** [comment]
Counterpoint: the reason we disagree isn't because we need to be "educated" that we're ["share-croppers"](https://github.com/soc/directories-rs/issues/47#issuecomment-478337412) who need to learn to be "well-behaved", but because we view the platform differently.

Many use macOS because it's the best way to get a unix with a good gui. Remember, with the command line on Mac we're talking about [an open-source *unix*](https://en.wikipedia.org/wiki/Darwin_%28operating_system%29) that effectively shares a userspace with Linux.

That shared GNU/Linux/POSIX userspace is what we think of when we use the command line on Mac. It's why we'd prefer to share system conventions with it when it makes sense to.

**Comment by XAMPPRocky** [comment]
This topic isn't being very productive, and I don't think it should be discussed further on this issue. We'd be better served by working towards one of other the aspects of configuration file management we've yet to address, rather than circling around a difference in opinion.

**Comment by BurntSushi** [comment]
While this may not be satisfactory for a number of reasons, I'd like to say that ripgrep's approach to configuration files has generally side-stepped all of the issues mentioned here, and generally has received zero complaints from users. The trick is to opt out of encoding knowledge about xdg (or similar) at all in the first place. Namely, ripgrep requires users to set an environment variable that points to the location of the config file.

That location can be anywhere. The user can choose any arbitrary location. People like me who gave up caring about the cleanliness of their home directory can plop it at $HOME/.ripgreprc, while macOS users rebelling against their overlords can drop it into an xdg compatible location if it suits them.

As I said initially, this is not satisfactory for a number of reasons. One such example is that this really only applies to configuration. If your application needs to make use of cache directories, for example, then forcing the user to set an environment variable is probably bad UX and not something they'd expect. My approach also only works if the configuration file is entirely optional (which is honestly probably a generally good idea anyway).

Anywho, I don't mean to propose this as a panacea, but it's something that can work well in a specific circumstance for maintainers that don't want to be on the wrong side of an hord of users because your program put the config file in the wrong place.

**Comment by Alxandr** [comment]
I just published my first rust library on crates.io which is a config library that I plan to use in another application I'm developing. It's currently lacking quite a bit in documentation (but there is some), but it provides configuration (either configured by the application, or by convention) from multiple places (user folders, application folders, environment variables and cli arguments). Binding to structs is still not available though (but should be reasonably straight forward).

Sample usage looks like this:

```rust
#[derive(ClapConfig)]
pub struct Server {
  #[preftool_clap(help = "Server host")]
  host: String,

  #[preftool_clap(help = "Server port")]
  port: usize,

  #[preftool_clap(help = "Server certificate (enables TLS)")]
  cert: Option<String>,
}

fn main() -> std::io::Result<()> {
  let mut builder = TerminalLoggerBuilder::new();
  builder.level(Severity::Debug);
  builder.destination(Destination::Stdout);
  let log = builder.build().unwrap();

  let config = AppConfigBuilder::new("preftool-app-config-simple", Some(log))?
    .with_args::<Server>()
    .configure(|app| app.version("0.0.1"))
    .add_format(TomlConfigFormat)
    .build()?;

  println!("{:#?}", config);
  Ok(())
}
```

This will read config from toml files, environment variables, and application arguments.

**Comment by soc** [comment]
@Alxandr Very interesting library!

One note, please do not use `dirs-sys` directly – I updated the README to that effect: https://github.com/soc/dirs-sys-rs#compatibility

**Comment by Alxandr** [comment]
@soc issue is that neither one of those two provides me the information that I need though. I will probably have to go back to just having the functions embedded instead. I switched to `dirs-sys` to get away from having to do conditional dependencies based on OS. The only thing I use `home_dir`, `is_absolute_path` and 2 of the windows dirs.

**Comment by soc** [comment]
Feel free to use it at your own risk. I just wanted to make sure that the level of non-guarantees has been communicated. :-)

**Comment by Alxandr** [comment]
Fair enough. I would love a stable API with the low-level primitives I though ;-).

**Comment by cjbassi** [comment]
I just made/published [platform-dirs](https://github.com/cjbassi/platform-dirs-rs) which provides a simple API for retrieving platform dependent application directories and user directories. It also allows for specifying if an application is CLI based and uses the XDG specified application directories on macOS if so. There's some other differences relative to dirs-rs and directories-rs at the bottom of the readme too. It's still beta-level so the API may change and it may have bugs FYI. Any and all suggestions welcome!

**Comment by epage** [comment]
From Discord @mike7c2
```
It's pretty neat at the user end, I didn't have to do much at all to get it running, it does layering off cli args, config file args and environment args, it has some tools for other layers too but I didn't look at them

I really liked being able to define my config as a structure - that was easy and natural, really bolstered by the fact it populates comments from the struct into the CLI arg definitions and stuff. I do wonder if I should be going a step further and using a yaml/dhall schema to define my arguments but at the same time I don't think that added level of descriptiveness will get me anything

I didnt check in detail but it definitly pulled in quite a bit of stuff, exe size got big adding it but that isn't a real issue for me

The thing that bugged me most about it was not being able to make nested config fields, it needs to work with a flat structure so to work with it you need to flatten any nested keys

Overall I'm just trying to find something which works good enough I can build into a template I will build some GRPC services on top of
```

**Comment by runiq** [comment]
Reformatted for your reading pleasure:

> It's pretty neat at the user end, I didn't have to do much at all to get it running, it does layering off cli args, config file args and environment args, it has some tools for other layers too but I didn't look at them
>
> I really liked being able to define my config as a structure - that was easy and natural, really bolstered by the fact it populates comments from the struct into the CLI arg definitions and stuff. I do wonder if I should be going a step further and using a yaml/dhall schema to define my arguments but at the same time I don't think that added level of descriptiveness will get me anything
>
> I didnt check in detail but it definitly pulled in quite a bit of stuff, exe size got big adding it but that isn't a real issue for me
>
> The thing that bugged me most about it was not being able to make nested config fields, it needs to work with a flat structure so to work with it you need to flatten any nested keys
>
> Overall I'm just trying to find something which works good enough I can build into a template I will build some GRPC services on top of

**Comment by rcorre** [comment]
Apologies if this isn't the right place to ask, happy to move elsewhere.

Of the suggested crates at the top, two are deprecated, leaving [`directories`](https://docs.rs/directories/6.0.0/directories/struct.ProjectDirs.html). That crate has the following behavior:

- the XDG base directory and the XDG user directory specifications on Linux
- the Known Folder API on Windows
- the Standard Directories guidelines on macOS 

For cofiguration, that results in:

```rust
// Lin: /home/alice/.config/barapp
// Win: C:\Users\Alice\AppData\Roaming\Foo Corp\Bar App\config
// Mac: /Users/Alice/Library/Application Support/com.Foo-Corp.Bar-App
```

Is this actually the desired behavior for a CLI app? Most of the CLI apps I use (rust or otherwise) seem to use `.config` on Linux and Mac, and sometimes Windows too. This means that I can find my configs in the same place across every OS, and I can have a [config](https://git.sr.ht/~rrc/.config/tree) (dotfiles) repository that I can clone on every OS to share config.
I'd never think to look under `AppData/Roaming` or `Application Support` for CLI tool configurations, but maybe that's just because I'm primarily a Linux user.

I was curious if folks think rust CLI apps should follow the "standard" per-os directories, or try to use XDG everywhere?

**Comment by epage** [comment]
imo native paths should be used on Linux and Windows. Mac is a constant debate.

**Comment by rcorre** [comment]
I just found the [etcetera](https://github.com/lunacookies/etcetera) crate, which allows you to choose:

> base_strategy::choose_base_strategy & app_strategy::choose_app_strategy: Uses Windows on Windows & XDG everywhere else. This is used by most CLI tools & some GUI tools on each platform.
> base_strategy::choose_native_strategy & app_strategy::choose_native_strategy: Uses Windows on Windows, Apple on macOS/iOS, & XDG everywhere else. This is used by most GUI applications on each platform.

**Comment by CAD97** [comment]
One further small caveat on "which paths to use" is that especially for developer-aimed CLI tooling it can sometimes be desirable to use the XDG conventions instead of Known Folder conventions on Windows. Thankfully, we also have a *very* strong signal for when this is the case: when the XDG base environment variable is set.

Separately, you have the concept of "portable mode" installations that want to keep their files out of system-global locations, either so they can stay localized to a removable drive, for user-local installation, or so they can be sure they can modify it without admin permissions.

(was going through the various places this has been discussed, and didn't see this point raises anywhere else, so decided here was as good a place as any to put it)
